worlds/blasphemous/Rules.py

```diff
@@ -1,1479 +1,3333 @@
+                        self._blasphemous_can_enemy_bounce(logic, enemy) and \
+                            self._blasphemous_can_cross_gap(logic, player, 3))
+    
+    def _blasphemous_opened_mom_ladder(self, player):
+        return self.has("D04Z02S06[NW]", player) or \
+            self.has("D04Z02S06[NE]", player) or \
+                self.has("D04Z02S06[N]", player) or \
+                    self.has("D04Z02S06[S]", player)
+    
+    def _blasphemous_opened_tsc_gate(self, player):
+        return self.has("D05Z02S11[W]", player) or \
+            self.has("D05Z02S11[Cherubs]", player)
+    
+    def _blasphemous_opened_ar_ladder(self, player):
+        return self.has("D06Z01S23[Sword]", player) or \
+            self.has("D06Z01S23[E]", player) or \
+                self.has("D06Z01S23[S]", player) or \
+                    self.has("D06Z01S23[Cherubs]", player)
+    
+    def _blasphemous_broke_bottc_statue(self, player):
+        return self.has("D08Z01S02[NE]", player) or \
+            self.has("D08Z01S02[SE]", player)
+    
+    def _blasphemous_opened_wothp_gate(self, player):
+        return self.has("D09Z01S05[W]", player) or \
+            self.has("D09Z01S05[SE]", player) or \
+                self.has("D09Z01S05[NE]", player)
+
+    def _blasphemous_opened_botss_ladder(self, player):
+        return self.has("D17Z01S04[N]", player) or \
+            self.has("D17Z01S04[FrontR]", player)
+    
+    def _blasphemous_upwarp_skips_allowed(self, logic):
+        return True if logic.value >= 2 else False
+        
+    def _blasphemous_mourning_skips_allowed(self, logic):
+        return True if logic.value >= 2 else False
+        
+    def _blasphemous_enemy_skips_allowed(self, logic, enemy):
+        return True if logic.value >= 2 and enemy.value == 0 else False
+        
+    def _blasphemous_unknown_skips_allowed(self):
+        return False
+    
+    def _blasphemous_precise_skips_allowed(self):
+        return False
+
+    
+    def _blasphemous_can_beat_boss(self, boss: str, logic, player):
+        def has_boss_strength(name: str) -> bool:
+            silver: int = self.count("Quicksilver", player) if self.has("D01Z05S27[E]", player) else 0
+            flasks: int = self.count("Empty Bile Flask", player) if \
+                (self.has("D01Z05S18[E]", player) or self.has("D02Z02S09[E]", player) or \
+                self.has("D03Z02S14[E]", player) or self.has("D03Z03S03[SE]", player) or \
+                self.has("D04Z02S13[W]", player) or self.has("D05Z01S12[E]", player) or \
+                self.has("D20Z01S08[W]", player)) else 0
+
+            playerStrength: float = self.count("Life Upgrade", player) * 0.25 / 6 + \
+                self.count("Mea Culpa Upgrade", player) * 0.25 / 7 + self.count("Fervour Upgrade", player) * 0.20 / 6 \
+                + flasks * 0.15 / 8 + silver * 0.15 / 5
+            
+            bossStrength: float
+
+            if name == "warden":
```

Comment:
Feel like I'd almost always prefer this being a dictionary lookup. Don't know if it's a huge deal but it should be faster.

Reply:
Oh damn that didn't work. Still figuring out how to mark multiple lines. I meant the entire if-elif-elif-elif block that starts around here